execution-plane/arcon/src/streaming/task/filter.rs

```diff
@@ -29,60 +19,89 @@ where
 {
     pub fn new(
         udf: Arc<Module>,
-        in_channels: Vec<Channel>,
-        out_channels: Box<ChannelStrategy<IN>>,
     ) -> Self {
         let ctx = WeldContext::new(&udf.conf()).unwrap();
         Filter {
-            ctx: ComponentContext::new(),
-            _in_channels: in_channels,
-            out_channels,
             udf: udf.clone(),
             udf_ctx: ctx,
             metric: TaskMetric::new(),
+            _in: PhantomData,
         }
     }
 
-    fn handle_watermark(&mut self, w: Watermark) -> ArconResult<()> {
-        self.push_out(ArconEvent::Watermark(w))
-    }
-
-    fn handle_element(&mut self, element: &ArconElement<IN>) -> ArconResult<()> {
-        if let Ok(result) = self.run_udf(&(element.data)) {
-            // Check WeldBool
-            // On true, then pass along the element
-            if result == 1 {
-                let _ = self.push_out(ArconEvent::Element(*element));
-            }
-        } else {
-            // Just report the error for now...
-            error!(self.ctx.log(), "Failed to execute UDF...",);
-        }
-        Ok(())
-    }
-
     fn run_udf(&mut self, event: &IN) -> ArconResult<WeldBool> {
         let run: ModuleRun<WeldBool> = self.udf.run(event, &mut self.udf_ctx)?;
         let ns = run.1;
         self.metric.update_avg(ns);
         Ok(run.0)
     }
+}
 
-    fn push_out(&mut self, event: ArconEvent<IN>) -> ArconResult<()> {
-        let _ = self.out_channels.output(event, &self.ctx.system())?;
-        Ok(())
+impl<IN> Task<IN, IN> for Filter<IN>
+where
+    IN: 'static + ArconType,
+{
+    fn handle_watermark(&mut self, _w: Watermark) -> ArconResult<Vec<ArconEvent<IN>>> {
+        Ok(Vec::new())
+    }
+
+    fn handle_element(&mut self, element: ArconElement<IN>) -> ArconResult<Vec<ArconEvent<IN>>> {
```

Comment:
utilising the ArconResult

`let result = self.run_udf(&(element.data))?;`
`if result == 1 {
   Ok(vec!(ArconEvent::Element(element))
} else {
   Ok(Vec::new())
}
`